kernel/environ.ml

```diff
@@ -893,3 +893,70 @@ struct
   let compare _env c1 c2 = GlobRef.CanOrd.compare c1 c2
   let hash _env c = GlobRef.CanOrd.hash c
 end
+
+(** Renaming *)
+
+let rename old = function
+  | Anonymous -> old
+  | Name _ as x -> {old with Context.binder_name=x}
+
+let rec rename_type c = function
+  | [] -> c
+  | name :: rest as renamings ->
+    match kind c with
+    | Prod (old, a, b) -> mkProd (rename old name, a, rename_type b rest)
+    | LetIn (x,a,b,c) -> mkLetIn (x,a,b, rename_type c renamings)
+    | Cast (c,_,_) -> rename_type c renamings (* why not preserve the cast? *)
+    | _ -> c
+
+let mod_globals env f = {env with env_globals = f env.env_globals}
+
+let rename_constant c names env =
+  mod_globals env (fun globals ->
+      {globals with
+       Globals.constants =
+         globals.Globals.constants |>
+         Cmap_env.modify c (fun _ (cb,info) ->
+             {cb with const_type = rename_type cb.const_type names}, info)})
+
+let mod_inductive (mind,i) env f =
+  mod_globals env (fun globals ->
+      {globals with
+       Globals.inductives =
+         globals.Globals.inductives |>
+         Mindmap_env.modify mind (fun _ (mb,info) ->
+             {mb with
+              mind_packets = Array.mapi (fun j packet -> if i = j then f packet else packet)
+                  mb.mind_packets}, info)})
+
+let rec rename_ctx ctx names = match ctx, names with
+  | _, [] -> ctx
+  | (LocalDef _ as elt) :: ctx, _ -> elt :: (rename_ctx ctx names)
+  | (LocalAssum _ as elt)  :: ctx, Anonymous :: names -> elt :: (rename_ctx ctx names)
+  | LocalAssum (x,t) :: ctx, (Name _ as y) :: names ->
+    LocalAssum (rename x y,t) :: (rename_ctx ctx names)
+  | [], _ :: _ -> []
+(* TODO change semantics to error when too many renames *)
+
+(* This works because parameters are included in the inductive_body
+   components. If that changes we will need some other method. *)
+let rename_ind_type names packet =
+  match packet.mind_arity with
+  | RegularArity arity ->
+    let mind_user_arity = rename_type arity.mind_user_arity names in
+    { packet with mind_arity = RegularArity {mind_sort = arity.mind_sort; mind_user_arity} }
+  | TemplateArity _ -> { packet with mind_arity_ctxt = rename_ctx packet.mind_arity_ctxt names }
+
+let rename_ind ind names env =
+  mod_inductive ind env (fun packet -> rename_ind_type names packet)
+
+let rename_ctor (ind,c) names env =
+  mod_inductive ind env (fun packet ->
+      {packet with mind_user_lc = Array.mapi (fun j t ->
```

Comment:
What about the other cached data like expanded contexts and the like? We also rely on them for some uses.

Reply:
I changed so that mind_arity_ctxt is always updated (not just when template) and mind_nf_lc is updated
mind_params_ctxt is shared for the whole block so doesn't make sense to update when part of the block gets renamed.